<a href="https://colab.research.google.com/github/Hernan4444/INF3842-2025-2/blob/main/Codigos/%5BMIA%5D_Taller_No_Evaluado_Matplotlib%2C_Seaborn_y_Cr%C3%ADtica_a_visualizaci%C3%B3n_%5BEnunciado%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Altair, Matplotlib/Seaborn y crítica a visualización

## Taller **No evaluado**.

**Profesor:** Hernán Valdivieso.

**Ayudantes**: Francisca Ibarra y Pablo Araneda.

-----

El siguiente _jupyter notebook_ tiene como fín aplicar conocimiento de Matplotlib/Seaborn. Para esto se incluyen 3 actividades donde utilizará las librerías Matplotlib/Seaborn vistas en clases. Finalmente, tendrá una actividad donde utilizará los conceptos de Principios de Diseño para criticar dos visualizaciones.

Para facilitar la instalación de esta librería, este archivo fue entregado en un [Google Colab](https://colab.research.google.com/) para poder ejecutarlo en los servidores de Google, los cuales ya disponen de todos los recursos necesarios para ejecutar los códigos que implementarás y únicamente requerirá de internet para trabajar con este archivo.


Este  _jupyter notebook_ se compone de 5 actividades, las cuales se dividen en 2 partes:

2. Se entrega un _dataset_ y se pide que construyan una visualización con Matplotlib y/o Seaborn respetando ciertas condiciones (4 visualizaciones distintas).
3. Criticar una visualización y proponer una nueva versión aplicando los conceptos de Principios de Diseño. (2 visualización).

## Descargar _datasets_

Antes de empezar. Vamos a descargar un [ZIP con los _datasets_](https://drive.google.com/file/d/1YIf5FJqnOR-bgwuTJA5k9zkIqJUn7DNb/view?usp=share_link) necesarios en todo este taller. Para esto ocuparemos una función encargada de bajar un archivo desde google drive.

In [ ]:
import pandas as pd
import requests

def download_file_without_authenticate(id, destination):
    def get_confirm_token(response):
        for key, value in response.cookies.items():
            if key.startswith("download_warning"):
                return value

    URL = "https://docs.google.com/uc?export=download"
    response = requests.get(URL, params={"id": id, "confirm": 1}, stream=True)

    CHUNK_SIZE = 32768
    with open(destination, "wb") as f:
        for i, chunk in enumerate(response.iter_content(CHUNK_SIZE)):
            if chunk:  # filter out keep-alive new chunks
                f.write(chunk)

    return None

download_file_without_authenticate("1ZO8NIha5XLoQM2beqFUuiYbjP5VS7ID_", "taller-matplotlib-critica.zip")
!unzip -o -q taller-matplotlib-critica.zip

Verificamos qué archivos 'csv' tenemos descargados

In [ ]:
!ls

'Airline Survey.csv'	    'Models Error.csv'	 taller-matplotlib-critica.zip
'Fashion MNIST Result.csv'   sample_data


# Parte 1 - Uso de Matplotlib/Seaborn

Esta parte del práctico tiene por propósito poner en práctica la implementación de visualizaciones mediante programación. Se establecen tres casos con datasets y codificaciones objetivo para que implementes. **Solo debes utilizar las librerías pandas, matplotlib y seaborn.**

Se espera que el gráfico cumpla con todas las condiciones solicitadas, es decir, colores, títulos, nombre de ejes, etc. **Si un elemento no es mencionado en el enunciado, puedo hacer el supuesto que desea**. Por ejemplo:

1. Si no se solicita un nombre al gráfico, queda a su criterio si le incluye un nombre y cuál sea dicho nombre.

2. Si un gráfico no indica qué paleta o color utilizar, pueden ocupar cualquier color.

## Actividad 1

Un _dataset_ bien conocido es el [Fashion MNIST](https://www.kaggle.com/datasets/zalando-research/fashionmnist), el cual se compone de imágenes de ropa. Se entrenó un Multilayer Perceptron (MLP) que recibió la información de cada pixel de la imagen y clasificaba la ropa. Se provee un dataset tabular ([Fashion MNIST Result.csv](https://drive.google.com/file/d/1-W2xRuP6TOXMCNAD1OmsRlkffEvb2J3z/view?usp=share_link)) que especifica la cantidad de respuestas dadas por el modelo en función de la clase real y la clase predicha. Use este _dataset_ para
visualizar esta información como una matriz de confusión(mapa de calor o _heatmap_).

Además, **debe** realizar las siguientes personalizaciones:

1. Agregar anotaciones sobre las celdas para leer los valores exactos

2. Utilizar la paleta de colores secuencial llamada `Oranges` (si quiere conocer más paletas, visite [este link](https://seaborn.pydata.org/tutorial/color_palettes.html)).

A continuación se muestra un ejemplo de como debe verse el gráfico solicitado.

<center>

![AC04](https://res.cloudinary.com/hernan4444/image/upload/v1652804896/Diplomado/ML/Altair_Matplotlib_AC04.png)

</center>

In [ ]:
# Código para asegurar un buen tamaño del gráfico
import matplotlib.pyplot as plt
plt.rcParams['figure.figsize'] = (15.0, 8.0) # permite específicar las dimensiones del gráfico creado

# Abrir dataset
dataset = pd.read_csv('Fashion MNIST Result.csv')
dataset.head()

True Label  Predictions
0         Coat        Dress
1         Coat        Shirt
2  T-shirt/top  T-shirt/top
3         Coat     Pullover
4   Ankle boot   Ankle boot

In [ ]:
# Completar con su respuesta


## Actividad 2

Una empresa de aerolíneas entrevisto a una gran cantidad de pasajeros para recopilar información sobre su vuelo (tiempo de espera, tipo de cliente, sector del asiento, género, etc). Además, se le solicitó que califiquen si el servicio entregado durante el vuelo fue satisfactorio o no. Con dicha información, la aerolínea busca construir un modelo que sea capaz de predecir la satisfacción de un cliente según su información previa. Se provee un dataset tabular ([Airline Survey.csv](https://drive.google.com/file/d/1X2_X2hSFoe86aIicb_9M9Pkcn5EZkLwQ/view?usp=share_link)) con la información recopilada.


### Caso 1


Antes de comenzar con el proceso de ML, la empresa desea asegurar que los datos recopilados no presenten un sesgos en relación al género. En particular, que los datos recopilados no vengan únicamente de hombres o de mujeres, sino que esté balanceado. Para observar esta proporción de géneros, se solicita que implemente una visualización con **gráfico de tipo torta**. El gráfico debe:

- Mostrar la proporción de encuestados por género. Para mostrar las proporciones, investigue sobre el argumento `autopct` de plt.pie.

- Tener como título "Proporción de encuestados por género".

A continuación se muestra un ejemplo de como debe verse el gráfico solicitado.

<center>

![caso_2_posible_1](https://res.cloudinary.com/hernan4444/image/upload/v1652808418/Diplomado/ML/Altair_Matplotlib_AC05.png)

</center>

In [ ]:
# Código para descargar y abrir dataset
import pandas as pd

df_airline = pd.read_csv("Airline Survey.csv", usecols=["id", "Gender", "satisfaction"])
df_airline.head()

id  Gender             satisfaction
0   70172    Male  neutral or dissatisfied
1    5047    Male  neutral or dissatisfied
2  110028  Female                satisfied
3   24026  Female  neutral or dissatisfied
4  119299    Male                satisfied

In [ ]:
# Completar con su respuesta



### Caso 2

Aparte de evaluar el segso de sus datos, la empresa mostró un fuerte interes en querer premiar a sus clientes más leales, y buscar cómo llegar a sus clientes no leales y convertirlos en leales. Por este motivo, necesita caracterizar un poco a estos 2 tipos de clientes. En particular, busca crear 2 visualizaciones que permitan conocer cómo distribuye la distancia de vuelo y la edad según el tipo de cliente. De este modo, determinar mejor qué tipo de acciones tomar.

La visualización a confeccionar corresponde a 2 gráficos de múltiples violines. En cada uno de estos se mostrará la distribución del tipo de cliente según una característica: el gráfico de la izquierda será según la distancia recorrida en los vuelos, mientras que el gráfico de la derecha será según su edad. En ambos casos, se espera los gráficos de violín incluyan el rango interquartile y la mediana.

A continuación se muestra un ejemplo de como debe verse el gráfico solicitado.

<center>

![AC2_CASO2](https://res.cloudinary.com/hernan4444/image/upload/v1684190433/Diplomado/ML/Matplotlib_AC02_Caso_2.png)

</center>

In [ ]:
# Código abrir dataset
import pandas as pd
import matplotlib.pyplot as plt

df_airline = pd.read_csv("Airline Survey.csv",
                         usecols=["Customer Type", "Flight Distance", "Age"])
df_airline.head()

Customer Type  Age  Flight Distance
0     Loyal Customer   13              460
1  disloyal Customer   25              235
2     Loyal Customer   26             1142
3     Loyal Customer   25              562
4     Loyal Customer   61              214

In [ ]:
# Completar con su respuesta



## Actividad 3

Una investigación realizada hace un par de años consitió en desarrollar un modelo capaz de clasificar noticias en función de su temática (economica, politica, violencia, moral). Un experimento consistió en desarrollar varios modelos de _deep learning_, entrenados con un conjunto de noticias y testeados con otro. Se provee un _dataset_ tabular ([Models Error.csv](https://drive.google.com/file/d/1Jgiyan9T734XXcOs4nC6Ha23SUwG2u1J/view?usp=share_link)) que especifica el error que tuvieron los modelo en las 12 épocas que fue entrenado cada uno. La idea es generar una visualización que permita observar la evaluación del error en cada set de datos para identificar si algún modelo fue exitoso o no.

Use este _dataset_ para construir un **_small multiple_**. Este gráfico debe cumplir las siguientes condiciones:
- Cada gráfico será un gráfico de multilíneas (con 2 líneas por gráfico).
- Cada gráfico corresponde a un modelo distinto.
- El color de cada línea en el gráfico será definido según el set de datos ocupados para obtener el error (`train` o `test`).
- El eje X será la época y el eje Y será el error del modelo en dicha época.


A continuación se muestra un ejemplo de cómo debe verse el gráfico solicitado.

<center>

![AC06](https://res.cloudinary.com/hernan4444/image/upload/v1652812329/Diplomado/ML/Altair_Matplotlib_AC06.png?1)

</center>




In [ ]:
# Abrir dataset
dataset_ml = pd.read_csv('Models Error.csv')
dataset_ml.head(14)

model  epoch     train      test
0   Modelo 1      0  0.984718  0.819433
1   Modelo 1      1  0.390081  0.549909
2   Modelo 1      2  0.102278  0.632147
3   Modelo 1      3  0.009471  0.565434
4   Modelo 1      4  0.007925  0.385672
5   Modelo 1      5  0.008474  0.236328
6   Modelo 1      6  0.005226  0.469562
7   Modelo 1      7  0.002951  0.730432
8   Modelo 1      8  0.001438  0.932933
9   Modelo 1      9  0.000714  0.964245
10  Modelo 1     10  0.000604  0.851745
11  Modelo 1     11  0.000440  0.861950
12  Modelo 2      0  0.850002  0.757395
13  Modelo 2      1  0.414887  0.344107

In [ ]:
# Completar con su respuesta


# Parte 2 - Criticar visualizaciones

## Actividad 4 - Pandemias más mortales

El 2020 se publicó la siguiente infografía para visualizar la historia de las pandemias más mortales. En particular, para permitir al usuario dimensionar la cantidad de muertes causadas por cada una y en qué año ocurrieron.

![AC05](https://www.visualcapitalist.com/wp-content/uploads/2020/03/HoP-Main-Graphic-03052022.jpg)

Fuente: https://www.visualcapitalist.com/history-of-pandemics-deadliest/

En esta infografía:
- El tamaño de cada pandemia representa la cantidad de fallecimientos provocados por esta.
- Su posición indica el año donde empezó.
- El color es utilizado con fines estéticos para distinguir cada pandemia.


**Pregunta 1** - ¿Cuáles son 3 **principales** principios de diseño que no cumple esta visualización? Justifique su respuesta en un máximo de **200 palabras**.

Respuesta:
- Completar


- Completar

- Completar


**Pregunta 2** -  Proponga una nueva visualización, aplicando los cambios necesarios a la visualización original, para cumplir con los principios de diseño y poder solucionar las **tareas visuales** ( permitir al usuario dimensionar la cantidad de muertes causadas por estas y en qué año ocurrieron). En esta ocasión, **no es necesario** que su solución debe ser la más óptima para cumplir las tareas visuales. Debe describir adecuadamente los cambios propuestos, puede incluir fotos de visualización en internet para apoyar su descripción. La descripción debe tener un máximo de **200 palabras**.


Respuesta:

Completar

## Actividad 5 - Similitudes de Lenguajes


El 2012 se publicó la siguiente visualización para observar las similitudes de los lenguajes en el mundo. En particular, para permitir al usuario identificar cuales lenguajes son similares según una métrica que mide diferencia de palabras (distancia de _Levenshtein_).


![AC06](https://res.cloudinary.com/hernan4444/image/upload/v1684177382/T5_2023_1_Caso2.jpg)

Fuente: https://hernan4444.github.io/Language-Network/ *

\* La página original es [esta](http://languagenetwork.cotrino.com/) pero ahí no funciona la visualización.

En esta visualización interactiva:
- Cada nodo circular representa un lenguaje. Su tamaño es proporcional al uso de dicho lenguaje en los datos utilizados. El color no codifica ninguna información, solo fue utilizado con fines estéticos.
- Cada línea entre dos nodos representa la similitud de los lenguajes.
- La posición de cada nodo no representa ninguna información, esta fue generada por un algoritmo que intentó reducir la cantidad de enlaces que se cruzan.
- El saturación de cada línea codifica el porcentaje de similitud. Una línea más saturada, indica que hay una mayor similaridad.

Respecto a las interacciones posibles:

- Se puede utilizar el _mouse_ para hacer _drag & drop_ en la visualización. Esto permite mover un nodo y volver a aplicar el algoritmo que determine la posición de todos los demás nodos.

- Se puede hacer _hover_ sobre un nodo. Esto permite ver información más detallada de este lenguaje y la similitud con otros lenguajes.

- Inicialmente se muestran todos los nodos del _dataset_, y con un _checkbox_ es posible filtrar aquellos nodos que no presentan enlaces. Permitiendo ver únicamente los lenguajes que tengan conexión con otro lenguaje.

- Inicialmente la visualización filtra aquellas líneas que representan un porcentaje menor al 50% de similitud. Luego, mediante un selector en la barra izquierda, es posible alterar ese filtro en un rango de 33% a 85% de similaridad. Permitiendo aumentar y reducir la cantidad de líneas que se ven en la visualización.




**Pregunta 1** - ¿Cuáles son 3 **principales** principios de diseño que no cumple esta visualización? Justifique su respuesta en un máximo de **200 palabras**.

Respuesta:
- Completar

- Completar

- Completar


**Pregunta 2** -  Proponga una nueva visualización, aplicando los cambios necesarios a la visualización original, para cumplir con los principios de diseño y poder solucionar las **tareas visuales** (para permitir al usuario identificar cuales lenguajes son similares). Debe describir adecuadamente los cambios propuestos, puede incluir fotos de visualización en internet para apoyar su descripción. La descripción debe tener un máximo de **200 palabras**.

Respuesta:

Completar